File: vane-regions/src/main/java/org/oddlama/vane/regions/RegionEconomyDelegate.java
package org.oddlama.vane.regions;

import net.milkbowl.vault.economy.Economy;
import net.milkbowl.vault.economy.EconomyResponse;
import org.bukkit.OfflinePlayer;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.RegisteredServiceProvider;

public class RegionEconomyDelegate {

    public Regions parent;
    public Economy economy;

    public RegionEconomyDelegate(final Regions parent) {
        this.parent = parent;
    }

    public Regions getModule() {
        return parent;
    }

    public boolean setup(final Plugin plugin) {
        RegisteredServiceProvider<Economy> rsp = getModule()
            .getServer()
            .getServicesManager()
            .getRegistration(Economy.class);
        if (rsp == null) {
            getModule().getLog().severe(
                    "Economy was selected as the currency provider, but no Economy service provider is registered via VaultAPI! Falling back to material currency."
                );
            return false;
        }

        economy = rsp.getProvider();
        return true;
    }

    public boolean has(final OfflinePlayer player, double amount) {
        return economy.has(player, amount);
    }

    public EconomyResponse withdraw(final OfflinePlayer player, double amount) {
        return economy.withdrawPlayer(player, amount);
    }

    public EconomyResponse deposit(final OfflinePlayer player, double amount) {
        return economy.depositPlayer(player, amount);
    }

    public String currencyNamePlural() {
        return economy.currencyNamePlural();
    }

    public int fractionalDigits() {
        return economy.fractionalDigits();
    }
}

File: vane-regions/src/main/kotlin/org/oddlama/vane/regions/RegionEconomyDelegate.kt
package org.oddlama.vane.regions

import net.milkbowl.vault.economy.Economy
import net.milkbowl.vault.economy.EconomyResponse
import org.bukkit.OfflinePlayer
import org.bukkit.plugin.Plugin

/**
 * Thin wrapper around Vault economy APIs used by the regions module.
 */
class RegionEconomyDelegate(private val module: Regions) {
    /**
     * Economy service provider resolved from Vault.
     */
    private var economy: Economy? = null
    /**
     * Non-null accessor for the resolved economy service.
     */
    private val activeEconomy: Economy
        get() = requireNotNull(economy) { "Economy delegate used before setup() completed" }

    /**
     * Resolves and stores the active Vault economy provider.
     *
     * @return `true` when setup succeeds.
     */
    fun setup(plugin: Plugin?): Boolean {
        if (plugin == null) {
            module.log.severe(
                "Economy was selected as the currency provider, but the Vault plugin wasn't found! Falling back to material currency."
            )
            return false
        }
        val rsp = module
            .server
            .servicesManager
            .getRegistration<Economy?>(Economy::class.java)
        if (rsp == null) {
            module.log.severe(
                "Economy was selected as the currency provider, but no Economy service provider is registered via VaultAPI! Falling back to material currency."
            )
            return false
        }

        economy = rsp.provider
        return true
    }

    /**
     * Returns whether the player has at least the requested balance.
     */
    fun has(player: OfflinePlayer?, amount: Double): Boolean = activeEconomy.has(player, amount)

    /**
     * Withdraws the requested amount from a player account.
     */
    fun withdraw(player: OfflinePlayer?, amount: Double): EconomyResponse = activeEconomy.withdrawPlayer(player, amount)

    /**
     * Deposits the requested amount into a player account.
     */
    fun deposit(player: OfflinePlayer?, amount: Double): EconomyResponse = activeEconomy.depositPlayer(player, amount)

    /**
     * Returns the pluralized currency label from the active provider.
     */
    fun currencyNamePlural(): String = activeEconomy.currencyNamePlural()

    /**
     * Returns the number of supported fractional digits in this economy.
     */
    fun fractionalDigits(): Int = activeEconomy.fractionalDigits()
}
